doc/kernel/usermode/kernelobjects.rst

```diff
@@ -0,0 +1,226 @@
+* An extra data field. This is currently used for thread stack objects
+  to denote how large the stack is, and for thread objects to indicate
+  the thread's index in kernel object permission bitfields.
+
+Supervisor Thread Access Permission
+===================================
+
+Supervisor threads can access any kernel object. However, permissions for
+supervisor threads are still tracked for two reasons:
+
+* If a supervisor thread calls :cpp:func:`k_thread_user_mode_enter()`, the
+  thread will then run in user mode with any permissions it had been granted
+  (in many cases, by itself) when it was a supervisor thread.
+
+* If a supervisor thread creates a user thread with the
+  :c:macro:`K_INHERIT_PERMS` option, the child thread will be granted the
+  same permissions as the parent thread, except the parent thread object.
+
+User Thread Access Permission
+=============================
+
+By default, when a user thread is created, it will only have access permissions
+on its own thread object. Other kernel objects by default are not usable.
+Access to them needs to be explcitly or implcitly granted. There are several
+ways to do this.
+
+* If a thread is created with the :c:macro:`K_INHERIT_PERMS`, that thread
+  will inherit all the permissions of the parent thread, except the parent
+  thread object.
+
+* A thread that has permission on an object, or is running in supervisor mode,
+  may grant permission on an object to another thread via the
+  :c:func:`k_object_access_grant()` API. The convenience function
+  :c:func:`k_thread_access_grant()` may also be used, which accepts a
+  NULL-terminated list of kernel objects and calls
+  :c:func:`k_object_access_grant()` on each of them. The thread being granted
+  permission, or the object whose access is being granted, do not need to be in
+  an initialized state. If the caller is from user mode, the caller must have
+  permissions on both the kernel object and the target thread object.
+
+* Supervisor threads may declare a particular kernel object to be a public
+  object, usable by all current and future threads with the
+  :c:func:`k_object_access_all_grant()` API. You must assume that any
+  untrusted or exploited code will then be able to access the object. Use
+  this API with caution!
+
+* If a thread was declared statically with :c:macro:`K_THREAD_DEFINE()`,
+  then the :c:macro:`K_THREAD_ACCESS_GRANT()` may be used to grant that thread
+  access to a set of kernel objects at boot time.
+
+Once a thread has been granted access to an object, such access may be
+removed with the :c:func:`k_object_access_revoke()` API. User threads using
+this API must have permission on both the object in question, and the thread
+object that is having access revoked.
+
+Initialization State
+====================
+
+Most operations on kernel objects will fail if the object is considered to be
+in an unitialized state. The appropriate init function for the object must
```

Comment:
"uninitialized"

Reply:
i'm terrible at typing this word for some reason :)